src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs

```diff
@@ -161,51 +169,76 @@ private PackageVerificationResult VerifyValidityAndTrust(Signature signature, Si
                         {
                             foreach (var message in messages)
                             {
-                                issues.Add(SignatureLog.Issue(treatIssuesAsErrors, NuGetLogCode.NU3021, message));
+                                issues.Add(SignatureLog.Issue(treatIssueAsError, NuGetLogCode.NU3012, message));
                             }
+
                             chainBuildingHasIssues = true;
                         }
 
+                        // Debug log any errors
+                        issues.Add(SignatureLog.DebugLog(
+                            string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.ErrorInvalidCertificateChain,
+                                string.Join(", ", chainStatuses.Select(x => x.Status.ToString())))));
+
                         // For all the special cases, chain status list only has unique elements for each chain status flag present
                         // therefore if we are checking for one specific chain status we can use the first of the returned list
                         // if we are combining checks for more than one, then we have to use the whole list.
                         IEnumerable<X509ChainStatus> chainStatus = null;
                         if (CertificateChainUtility.ChainStatusListIncludesStatus(chainStatuses, X509ChainStatusFlags.Revoked, out chainStatus))
                         {
                             var status = chainStatus.First();
-                            issues.Add(SignatureLog.Issue(true, NuGetLogCode.NU3021, status.StatusInformation));
+
+                            issues.Add(SignatureLog.Error(NuGetLogCode.NU3012, status.StatusInformation));
+
                             return SignatureVerificationStatus.Invalid;
                         }
 
+                        if (isSelfSignedCertificate &&
+                            CertificateChainUtility.TryGetStatusMessage(chainStatuses, X509ChainStatusFlags.UntrustedRoot, out messages))
+                        {
+                            issues.Add(SignatureLog.Issue(!allowUntrustedSelfSignedCertificate, NuGetLogCode.NU3018, messages.First()));
+
+                            if (!chainBuildingHasIssues && allowUntrustedSelfSignedCertificate)
+                            {
+                                return SignatureVerificationStatus.Trusted;
+                            }
+                        }
+
                         const X509ChainStatusFlags RevocationStatusFlags = X509ChainStatusFlags.RevocationStatusUnknown | X509ChainStatusFlags.OfflineRevocation;
                         if (CertificateChainUtility.TryGetStatusMessage(chainStatuses, RevocationStatusFlags, out messages))
                         {
-                            if (treatIssuesAsErrors)
+                            foreach (var message in messages)
```

Comment:
The reason for this if statement was to avoid warning in the VS client case, where this operation might be ran for multiple packages in a restore offline scenario, and the user shouldn't be filled with warnings in that case.